### PR TITLE
POS-tagging example

### PR DESCRIPTION
**Pull request by unnonouno**
This is an example of CRF1d function. `nltk` is required.
Please merge #1451 first.

### CODE REVIEW

**Comment by rezoo** [comment]
Sorry for the late review. Now I would like to see your PR.

**Comment by rezoo** [comment]
That is a good example when a novice NLP guy wants to implement POS in chainer.
The most critical issue is the lack of ``README.md``. I think it is better to add it in ``/chainer/examples/pos``, and describe the abstract of this script and requirements (``nltk``?).

File: examples/pos/postagging.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+    model = CRF(len(vocab), len(pos_vocab))
+    model.to_gpu()
+    xp = cuda.cupy
+    optimizer = O.Adam()
+    optimizer.setup(model)
+    #opt.add_hook(chainer.optimizer.WeightDecay(0.1))
+
+    train_data = data[len(data) // 4:]
+    test_data = data[:len(data) // 4]
```
Comment (rezoo) [inline examples/pos/postagging.py]: Since this is an example, please use the following instead of ``len(data) // 4``.
```python
n_test = len(data) // 4
train_data, test_data = data[n_test:], data[:n_test]
```

File: examples/pos/postagging.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+        data.append((xs, ys))
+
+    print('# of sentences: {}'.format(len(data)))
+    print('# of words: {}'.format(len(vocab)))
+    print('# of pos: {}'.format(len(pos_vocab)))
+
+    model = CRF(len(vocab), len(pos_vocab))
+    model.to_gpu()
+    xp = cuda.cupy
+    optimizer = O.Adam()
```
Comment (rezoo) [inline examples/pos/postagging.py]: Since ``chainer.optimizers`` is only used in that, please explicitly use `` chainer.optimizers.Adam()`` instead.

**Comment by unnonouno** [comment]
OK, thank you very much. And i'll also discuss if this example is enough or not for an example of sequential labeling with other NLPer.

**Comment by beam2d** [comment]
I read the MicroAverage implementation. I think it looks good to put into the main chainer package rather than the example, though it would need some design reviews. For example, I'm wondering if modifying the log report directly is a good way of reporting the statistics, or we should use reporter here.

File: examples/pos/postagging.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+    updater = training.StandardUpdater(
+        train_iter, optimizer, converter=convert, device=args.gpu)
+    trainer = training.Trainer(updater, (args.epoch, 'epoch'), out=args.out)
+
+    trainer.extend(extensions.Evaluator(
+        test_iter, model, device=args.gpu, converter=convert))
+    trainer.extend(extensions.LogReport())
+
+    trainer.extend(extensions.MicroAverage(
+        'main/correct', 'main/total', 'main/accuracy'))
```
Comment (rezoo) [inline examples/pos/postagging.py]: It seems that `main/accuracy` is undefined (at least in my environment it does not output `main/accuracy`).

File: examples/pos/postagging.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+    trainer.extend(extensions.MicroAverage(
+        'validation/main/correct', 'validation/main/total',
+        'validation/main/accuracy'))
```
Comment (rezoo) [inline examples/pos/postagging.py]: It seems that `validation/main/accuracy` is undefined.

File: examples/pos/postagging.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+
+    train_iter = chainer.iterators.SerialIterator(train_data, args.batchsize)
+    test_iter = chainer.iterators.SerialIterator(test_data, args.batchsize,
+                                                 repeat=False, shuffle=False)
+    updater = training.StandardUpdater(
+        train_iter, optimizer, converter=convert, device=args.gpu)
+    trainer = training.Trainer(updater, (args.epoch, 'epoch'), out=args.out)
+
+    trainer.extend(extensions.Evaluator(
+        test_iter, model, device=args.gpu, converter=convert))
```
Comment (rezoo) [inline examples/pos/postagging.py]: Is it better to add `trigger` options to all the extensions?

File: examples/pos/postagging.py (hunk 3 above)
Comment (unnonouno) [inline examples/pos/postagging.py]: In my environment, it works.

File: examples/pos/postagging.py (hunk 5 above)
Comment (unnonouno) [inline examples/pos/postagging.py]: Do you mean that it is better to add `trigger=(1, 'epoch')` explicitly?

**Comment by unnonouno** [comment]
I updated the branch based on your comment. In my environment, micro-average works correctly.

File: examples/pos/postagging.py (hunk 5 above)
Comment (rezoo) [inline examples/pos/postagging.py]: I think it is better to add it explicitly, or add a comment that it invokes per one epoch without trigger option

File: examples/pos/postagging.py (hunk 3 above)
Comment (rezoo) [inline examples/pos/postagging.py]: Ah sorry after restarting my environment it works.

**Comment by rezoo** [comment]
Thank you. Just two comments.
i) currently it prints the log per one epoch only, but I think it is better to print more frequently.
ii) it seems that ``len(data) / 10`` has not been fixed.

**Comment by beam2d** [comment]
@rezoo What is the status of this PR?

**Comment by rezoo** [comment]
Sorry for being so late... LGTM. Although it does not include a test for this example, I will add this in near future.
